### PR TITLE
Feature/issue-159 assertj integration

### LINKED ISSUE #159: Replace Hamcrest with AssertJ
**Is your feature request related to a problem? Please describe.**
At present, the codebase uses the standard Hamcrest matchers for test assertions. While largely adopted, this library hasn't an expressive API, which works fine for simple assertions but falls short for more complex cases making the code difficult to read, often forcing test authors to write helper methods or classes. Moreover, its extension mechanism is clunky, resulting in very few developers writing custom Matchers.

**Describe the solution you'd like**
A valid alternative exists in the [AssertJ](http://joel-costigliola.github.io/assertj/) library which has a fluent and very expressive API, and covers lots of common complex cases. It's fully compatible with Java 8+ and supports lambdas as assertion callbacks.
The API revolves around static methods `assertThat(obj)` that let you chain one or more assertions fluidly. To get a feeling of it check out the [core features highlight](http://joel-costigliola.github.io/assertj/assertj-core-features-highlight.html) page.

My proposal is to replace existing test assertions in BULL using AssertJ, which will provide a powerful and expressive API for complex assertions on objects/JavaBeans (there are a lot of them as test samples) without requiring extra code.

### PR DESCRIPTION
**Pull request by fborriello**
## Description

Replaces Hamcrest, Junit and Shazam libraries with AssertJ.
AssertJ now becomes the standard library to be used for all tests assertion.

Fixes #159 

## How Has This Been Tested?

- [X] All existing tests are working fine

## Checklist:

- [X] The branch follows the best practices naming convention:
     - Use grouping tokens (words) at the beginning of your branch names.
        * `feature`: Feature I'm adding or expanding
        * `bug`: Bug fix or experiment
        * `wip`: Work in progress; stuff I know won't be finished soon
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented on my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] Any dependent changes have been merged and published in downstream modules
- [X] My changes have no bad impacts on performances
- [X] My changes have been tested through Unit Tests and the overall coverage has not decreased. Check the coverall report for your branch: [here](https://coveralls.io/github/HotelsDotCom/bull)

### CODE REVIEW

File: bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java [hunk 1]
```diff
@@ -236,11 +231,11 @@ public void testTransformWorksProperlyWithMapContainingList() {
...
-        assertNotNull(actual);
-        assertEquals(actual.size(), sourceMap.size());
+        assertThat(actual).isNotNull();
+        assertThat(actual.size()).isEqualTo(sourceMap.size());
         for (Map.Entry<MutableToFooSimple, List> entry : actual.entrySet()) {
-            assertEquals(MutableToFooSimple.class, entry.getKey().getClass());
-            assertEquals(sampleList, entry.getValue());
+            assertThat(entry.getKey().getClass()).isEqualTo(MutableToFooSimple.class);
+            assertThat(entry.getValue()).isEqualTo(sampleList);
         }
```
Comment (polarene) [inline bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java:239]: GIVEN block could be simplified to:

    import static org.assertj.core.util.Lists.list;
    import static org.assertj.core.util.Maps.newHashMap;
    ...

    Map<FromFooSimple, List<String>> sourceMap = newHashMap(fromFooSimple, list(ITEM_1));

Assertions could be simplified to:

    assertThat(actual).allSatisfy((key, value) -> {
        assertThat(key).isInstanceOf(MutableToFooSimple.class);
        assertThat(value).containsOnly(ITEM_1);
    });

File: bull-bean-transformer/src/test/java/com/hotels/beans/BeanUtilsTest.java [hunk 2]
```diff
@@ -106,9 +103,9 @@ public void testGetTransformerFunctionWorksProperly(final String testCaseDescrip
                 .collect(Collectors.toList());
 
         //THEN
-        assertNotNull(transformerFunction);
+        assertThat(transformerFunction).isNotNull();
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/BeanUtilsTest.java:106]: I would leave as it is

File: bull-bean-transformer/src/test/java/com/hotels/beans/populator/ArrayPopulatorTest.java [hunk 3]
```diff
@@ -83,18 +82,18 @@ public void testGetPopulatedObjectWorksProperly(final Class<?> genericFieldType,
...
         } else if (genericFieldType == MixedToFooStaticField.class) {
             final MixedToFooStaticField[] expectedArray = (MixedToFooStaticField[]) array;
             final Object[] actualArray = (Object[]) actual;
             IntStream.range(0, expectedArray.length)
-                    .forEach(i -> assertEquals(expectedArray[i].getNormalField(), ((MixedToFooStaticField) actualArray[i]).getNormalField()));
+                    .forEach(i -> assertThat(((MixedToFooStaticField) actualArray[i]).getNormalField()).isEqualTo(expectedArray[i].getNormalField()));
         } else {
-            assertArrayEquals((Object[]) array, (Object[]) actual);
+            assertThat((Object[]) actual).isEqualTo(array);
         }
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/populator/ArrayPopulatorTest.java:97]: done

File: bull-bean-transformer/src/test/java/com/hotels/beans/populator/PopulatorFactoryTest.java [hunk 4]
```diff
@@ -88,9 +88,9 @@ public void testGetPopulatorReturnsTheExpectedResult(final Class type, final Cla
 
         // THEN
         final boolean isNonNullObjectExpected = nonNull(expectedResult);
-        assertEquals(populator.isPresent(), isNonNullObjectExpected);
+        assertThat(populator.isPresent()).isEqualTo(isNonNullObjectExpected);
         if (isNonNullObjectExpected) {
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/populator/PopulatorFactoryTest.java:92]: I know, but the purpose of this issue is to replace the test libraries with AssertJ, all the refactoring can be done in a separate issue.
If you don't mind I do prefer to apply the changes in scope with the issue

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 5]
```diff
@@ -145,7 +138,7 @@ public void testTransformationOnAnExistingDestinationWorksProperly() {
         underTest.setValidationEnabled(true).transform(fromFooSimple, immutableToFoo);
 
         //THEN
-        assertThat(immutableToFoo, sameBeanAs(fromFooSimple));
+        assertThat(immutableToFoo).usingRecursiveComparison().isEqualTo(fromFooSimple);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:141]: done

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 6]
```diff
@@ -171,9 +164,9 @@ public void testTransformationWithCompositeFieldNameMappingIsWorkingAsExpected(f
         ImmutableFlatToFoo actual = underTestMock.withFieldMapping(phoneNumbersMapping).transform(sourceObject, ImmutableFlatToFoo.class);
 
         //THEN
-        assertEquals(expectedName, actual.getName());
-        assertEquals(expectedId, actual.getId());
-        assertEquals(expectedPhoneNumbers, actual.getPhoneNumbers());
+        assertThat(actual.getName()).isEqualTo(expectedName);
+        assertThat(actual.getId()).isEqualTo(expectedId);
+        assertThat(actual.getPhoneNumbers()).isEqualTo(expectedPhoneNumbers);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:169]: done

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 7]
```diff
@@ -260,12 +253,12 @@ public void testImmutableBeanWithDifferentFieldNamesIsCorrectlyCopied() {
         ImmutableToFooDiffFields actual = beanTransformer.transform(fromFoo, ImmutableToFooDiffFields.class);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:253]: why?

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 8]
```diff
@@ -289,13 +282,13 @@ public void testImmutableBeanWithAdvancedFieldsIsCorrectlyCopied(final String te
         ImmutableToFooAdvFields actual = (ImmutableToFooAdvFields) beanTransformer.transform(sourceObject, targetObjectClass);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:282]: agree

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 9]
```diff
@@ -289,13 +282,13 @@ public void testImmutableBeanWithAdvancedFieldsIsCorrectlyCopied(final String te
...
-        assertEquals(sourceObject.getAge().get(), actual.getAge());
-        assertEquals(sourceObject.getClassType(), actual.getClassType());
-        assertEquals(sourceObject.getLocale(), actual.getLocale().getLanguage());
+        assertThat(actual).isNotNull();
+        assertThat(actual.getName().isPresent()).isEqualTo(isNameFieldEmpty);
+        sourceObject.getName().ifPresent(name -> assertThat(actual.getName().get()).isEqualTo(name));
+        assertThat(sourceObject.getAge().isPresent()).isTrue();
+        assertThat(actual.getAge()).isEqualTo(sourceObject.getAge().get());
+        assertThat(actual.getClassType()).isEqualTo(sourceObject.getClassType());
+        assertThat(actual.getLocale().getLanguage()).isEqualTo(sourceObject.getLocale());
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:291]: done

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 10]
```diff
@@ -430,9 +423,9 @@ public void testTransformationReturnsAMeaningfulException() {
         }
 
         //THEN
-        assertNotNull(raisedException);
-        assertEquals(InvalidBeanException.class, raisedException.getClass());
-        assertEquals(expectedExceptionMessage, raisedException.getMessage());
+        assertThat(raisedException).isNotNull();
+        assertThat(raisedException.getClass()).isEqualTo(InvalidBeanException.class);
+        assertThat(raisedException.getMessage()).isEqualTo(expectedExceptionMessage);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:428]: fantastic!

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java [hunk 11]
```diff
@@ -464,9 +457,9 @@ public void testFieldTransformationSkipWorksProperly() {
         ImmutableToFoo actual = underTest.transform(fromFoo, ImmutableToFoo.class);
 
         //THEN
-        assertEquals(fromFoo.getId(), actual.getId());
-        assertNull(actual.getName());
-        assertNull(actual.getNestedObject().getPhoneNumbers());
+        assertThat(actual.getId()).isEqualTo(fromFoo.getId());
+        assertThat(actual.getName()).isNull();
+        assertThat(actual.getNestedObject().getPhoneNumbers()).isNull();
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/ImmutableObjectTransformationTest.java:462]: done

File: bull-bean-transformer/src/test/java/com/hotels/beans/transformer/MixedObjectTransformationTest.java [hunk 12]
```diff
@@ -70,7 +64,7 @@ public void testMixedBeanIsCorrectlyCopied() {
         MixedToFoo actual = underTest.transform(fromFoo, MixedToFoo.class);
 
         //THEN
-        assertThat(actual, sameBeanAs(fromFoo));
+        assertThat(actual).usingRecursiveComparison().isEqualTo(fromFoo);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/transformer/MixedObjectTransformationTest.java:68]: done

File: bull-converter/src/test/java/com/hotels/beans/conversion/processor/impl/CharacterConversionTest.java [hunk 13]
```diff
@@ -211,6 +211,6 @@ public void testConvertBigDecimalShouldReturnProperResult() {
         char actual = underTest.convertBigDecimal().apply(BigDecimal.ZERO);
 
         // THEN
-        assertEquals(expectedValue, actual);
+        assertThat(actual).isEqualTo(expectedValue);
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-converter/src/test/java/com/hotels/beans/conversion/processor/impl/CharacterConversionTest.java:214]: not applicable to a char

File: bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java [hunk 14]
```diff
@@ -214,34 +206,32 @@ public void testTransformWorksProperlyWithTargetKeyAndElemType() {
...
-        assertNotNull(actual);
-        assertEquals(actual.size(), EXTREME_COMPLEX_MAP.size());
+        assertThat(actual).isNotNull();
+        assertThat(actual.size()).isEqualTo(EXTREME_COMPLEX_MAP.size());
         // check that the element has been converted
         for (Map.Entry<MutableToFooSimple, Map> entry : actual.entrySet()) {
-            assertEquals(MutableToFooSimple.class, entry.getKey().getClass());
+            assertThat(entry.getKey().getClass())
+                    .isEqualTo(MutableToFooSimple.class);
         }
```
Comment (polarene) [inline bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java:215]: Did you consider applying the simplification I proposed?
https://github.com/HotelsDotCom/bull/pull/161#discussion_r445456960

File: bull-converter/src/test/java/com/hotels/beans/conversion/processor/impl/CharacterConversionTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (polarene) [reply bull-converter/src/test/java/com/hotels/beans/conversion/processor/impl/CharacterConversionTest.java:214]: Sorry, I missed that!

File: bull-bean-transformer/src/test/java/com/hotels/beans/BeanUtilsTest.java [hunk 15]
```diff
@@ -105,7 +105,9 @@ public void testGetTransformerFunctionWorksProperly(final String testCaseDescrip
         //THEN
         assertThat(transformerFunction).isNotNull();
         IntStream.range(0, actual.size())
-                .forEach(i -> assertThat(actual.get(i)).isEqualToComparingFieldByField(fromFooSimpleList.get(i)).usingRecursiveComparison());
+                .forEach(i -> assertThat(actual.get(i))
+                        .isEqualToComparingFieldByField(fromFooSimpleList.get(i))
+                        .usingRecursiveComparison());
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/BeanUtilsTest.java]: cool, thank you

File: bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java (hunk 14 above)
Comment (fborriello) [reply bull-map-transformer/src/test/java/com/hotels/map/transformer/MapTransformerTest.java:215]: I missed it, thank you

File: bull-bean-transformer/src/test/java/com/hotels/beans/populator/ArrayPopulatorTest.java [hunk 16]
```diff
@@ -88,10 +87,9 @@ public void testGetPopulatedObjectWorksProperly(final Class<?> genericFieldType,
...
-            final MixedToFooStaticField[] expectedArray = (MixedToFooStaticField[]) array;
-            final Object[] actualArray = (Object[]) actual;
-            IntStream.range(0, expectedArray.length)
-                    .forEach(i -> assertThat(((MixedToFooStaticField) actualArray[i]).getNormalField()).isEqualTo(expectedArray[i].getNormalField()));
+            assertThat((Object[]) actual)
+                    .usingElementComparatorOnFields(NORMAL_FIELD)
+                    .isEqualTo(array);
         } else {
             assertThat((Object[]) actual).isEqualTo(array);
         }
```
[the remark this answers is not in the document]
Comment (fborriello) [reply bull-bean-transformer/src/test/java/com/hotels/beans/populator/ArrayPopulatorTest.java]: cool! thank you

**Pull request merged**